### ISSUE TITLE
[NX 20] Problems with empty workspace

### ISSUE THREAD

**Issue by gperdomor** [nrwl/nx#28322]
### Current Behavior

Hi everyone... First things first... Congrats for the new release 🥳 

Im creating this issue because I'm trying to play with Nx 20.0.0-rc.0 and I got several problems running `npx nx add @nx/nest`, `npx nx add @nx/next` and `npx nx add @nx/react`

Basically all commands return mostly the same: 
```
npx nx add @nx/react

✔ Installing @nx/react@20.0.0-rc.0...

 NX  Generating @nx/react:init


 NX   The "@nx/react" plugin doesn't yet support the existing TypeScript setup

We're working hard to support the existing TypeScript setup with the "@nx/react" plugin. We'll soon release a new version of Nx with support for it.

✖ Initializing @nx/react...

Error: Command failed: npx nx g @nx/react:init --keepExistingVersions --updatePackageScripts
    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at checkExecSyncError (node:child_process:890:11)
    at execSync (node:child_process:962:15)
    ...
    ...
    ...

 NX   Failed to initialize @nx/react. Please check the error above for more details.
```

### Creating Plugins
Also I ran `npx nx add @nx/plugin` which not throws errors during the install, but when I tried to create a new plugin I got the same error

```
Executing task: npx nx generate @nx/plugin:plugin --directory=plugin/my-plugin --no-interactive --dry-run 


 NX  Generating @nx/plugin:plugin


 NX   The "@nx/plugin:plugin" generator doesn't yet support the existing TypeScript setup

We're working hard to support the existing TypeScript setup with the "@nx/plugin:plugin" generator. We'll soon release a new version of Nx with support for it.
```

### Expected Behavior

Considering that not all plugins are supported with the new typescript plugin, maybe the plugin should not be included as default 🤔 

Also maybe it's a good idea include in the docs which plugins/executors/generators works with the new preset

### Steps to Reproduce

1. Create empty workspace
2. Execute `npx nx add @nx/next`

### Nx Report

 NX   Report complete - copy this into the issue template

Node           : 22.9.0
OS             : darwin-x64
Native Target  : x86_64-macos
npm            : 10.8.3

nx (global)    : 19.8.3
nx             : 20.0.0-rc.0
@nx/js         : 20.0.0-rc.0
@nx/jest       : 20.0.0-rc.0
@nx/eslint     : 20.0.0-rc.0
@nx/workspace  : 20.0.0-rc.0
@nx/devkit     : 20.0.0-rc.0
@nx/nest       : 20.0.0-rc.0
@nx/next       : 20.0.0-rc.0
@nx/node       : 20.0.0-rc.0
@nx/plugin     : 20.0.0-rc.0
@nx/react      : 20.0.0-rc.0
@nx/web        : 20.0.0-rc.0
@nx/webpack    : 20.0.0-rc.0
typescript     : 5.5.4
---------------------------------------
Registered Plugins:
@nx/js/typescript

### Package Manager Version

npm 10.8.3

### Operating System

- [x] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by Crewting-Niklas** [comment]
Same problem on my side. I used following commands to create a new workspace:

`npx create-nx-workspace --pm pnpm`
`pnpm i -D -w @nx/react`
`nx g @nx/react:application app --directory app`


**Console Report:**
 NX  Generating @nx/react:application

✔ Which stylesheet format would you like to use? · @emotion/styled
✔ Would you like to add React Router to this application? (y/N) · false
✔ Which E2E test runner would you like to use? · cypress
✔ Which bundler do you want to use to build the application? · vite

 NX   The "@nx/react:application" generator doesn't yet support the existing TypeScript setup

We're working hard to support the existing TypeScript setup with the "@nx/react:application" generator. We'll soon release a new version of Nx with support for it.

**Comment by chioarub** [comment]
Same problem when using @nx/angular.

**Comment by madelk** [comment]
Also getting a problem that might be related after going to NX 20
```
 NX   Cannot read properties of undefined (reading 'projects')

TypeError: Cannot read properties of undefined (reading 'projects')
    at getProjectConfiguration (/mypath/node_modules/.pnpm/@ns3+nx-core@5.4.0_nx@20.0.0_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc+helpers@0.5.11__@s_lglvwxjqsajdvzilv64lux6rua/node_modules/@ns3/nx-core/src/utils/get-project-configuration.js:6:30)
    at /mypath/node_modules/.pnpm/@ns3+nx-serverless@5.4.0_@babel+traverse@7.24.5_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc_2s7xprqvnfumnnfe7pzlv3jf5m/node_modules/@ns3/nx-serverless/src/executors/sls/executor.js:13:67
    at Generator.next (<anonymous>)
    at /mypath/node_modules/.pnpm/tslib@2.6.2/node_modules/tslib/tslib.js:169:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/mypath/node_modules/.pnpm/tslib@2.6.2/node_modules/tslib/tslib.js:165:16)
    at runExecutor (/mypath/node_modules/.pnpm/@ns3+nx-serverless@5.4.0_@babel+traverse@7.24.5_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc_2s7xprqvnfumnnfe7pzlv3jf5m/node_modules/@ns3/nx-serverless/src/executors/sls/executor.js:10:20)
    at runExecutorInternal (/mypath/node_modules/.pnpm/nx@20.0.0_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc+helpers@0.5.11__@swc+types@0.1.7_type_acywtu53f4pqbt6pniy2gi3ee4/node_modules/nx/src/command-line/run/run.js:98:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /mypath/node_modules/.pnpm/nx@20.0.0_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc+helpers@0.5.11__@swc+types@0.1.7_type_acywtu53f4pqbt6pniy2gi3ee4/node_modules/nx/src/command-line/run/run.js:176:44
```

I assumed it's related because in this function
```
function getProjectConfiguration(context) {
    return context.workspace.projects[context.projectName];
}
```
workspace is undefined

**Comment by baoletrangia** [comment]
same problem when adding `@nx/next` and `@nx/nest` to empty workspace

**Comment by leosvelperez** [comment]
Thanks for reporting this!

The new TS behavior is enabled by default when using the `ts` preset. As you've discovered, it's not yet compatible with most of the plugins. As of right now, it can be used with `@nx/js` projects. As the error message says, we're working to make all Nx first-party plugins compatible with it.

This doesn't mean you have to be blocked. As mentioned, this is only the behavior when using the `ts` preset. There are different presets for "empty" workspaces, and that's only one of them. As mentioned in [Nx v20 blog post announcement](https://nx.dev/blog/announcing-nx-20#typescript-project-references-for-monorepos), if you need to use other plugins while they don't support the new TS setup, please create an empty workspace with the `apps` preset: `npx -y create-nx-workspace@latest --preset=apps`.

**Comment by linktogo** [comment]
same issue when we do an migration command 

`
nx migrate latest --interactive

 >  NX   Its time to update Nx 🎉

   Your repository uses a higher version of Nx (18.3.3) than your global CLI version (16.10.0)
   For more information, see https://nx.dev/more-concepts/global-nx

Failed to install the latest version of the migration script. Using the current version.
yarn run v1.22.19
$ node --max-old-space-size=20384 ./node_modules/nx/bin/nx.js _migrate latest --interactive
✔ Would you like remote caching to make your build faster? · skip
Fetching meta data about packages.
It may take a few minutes.
Fetching nx@20.0.0
Fetching @nx/jest@20.0.0
Fetching @nx/js@20.0.0
Fetching @nx/web@20.0.0
Fetching @nrwl/web@20.0.0
Fetching @nx/eslint@20.0.0
Fetching @nx/cypress@20.0.0
Fetching @nx/storybook@20.0.0
Fetching @nx/eslint-plugin@20.0.0
Fetching @nx/node@20.0.0
Fetching @nx/workspace@20.0.0
Fetching @nx/nest@20.0.0
Fetching @nx/angular@20.0.0
Fetching @nx/webpack@20.0.0
Fetching @nx/jest@20.0.0

 NX   The migrate command failed.


 NX   Command failed: npm install @nrwl/web@20.0.0

npm error code ETARGET
npm error notarget No matching version found for @nrwl/web@20.0.0.
npm error notarget In most cases you or one of your dependencies are requesting
npm error notarget a package version that doesn't exist.
`

**Comment by leosvelperez** [comment]
@linktogo that's unrelated. There's no v20 for `@nrwl/web` package (or any `@nrwl/*` packages). This was announced in [the Nx v20 blog post announcement](https://nx.dev/blog/announcing-nx-20#nx-all-the-things). The migrations you should take care of it, so please report that in a separate issue with more detailed information on how to reproduce it.

To unblock yourself, you can swap it with `@nx/web` before migrating. You can see the [migration docs](https://nx.dev/deprecated/rescope) for more info.

**Comment by gperdomor** [comment]
@leosvelperez Hi buddy... Thanks for the reply... Considering that right now, most of the plugins are affected, can you share a roadmap about how that migration will be handled? Maybe a compatibility matrix can be a good option to help people to decide which preset is better for their projects

**Comment by eduard-malakhov** [comment]
> Thanks for reporting this!
> 
> The new TS behavior is enabled by default when using the `ts` preset. As you've discovered, it's not yet compatible with most of the plugins. As of right now, it can be used with `@nx/js` projects. As the error message says, we're working to make all Nx first-party plugins compatible with it.
> 
> This doesn't mean you have to be blocked. As mentioned, this is only the behavior when using the `ts` preset. There are different presets for "empty" workspaces, and that's only one of them. As mentioned in [Nx v20 blog post announcement](https://nx.dev/blog/announcing-nx-20#typescript-project-references-for-monorepos), if you need to use other plugins while they don't support the new TS setup, please create an empty workspace with the `apps` preset: `npx -y create-nx-workspace@latest --preset=apps`.

Thank you for your the explanation. What would be the correct way to disable the `ts` preset in an existing workspace? I'd rather avoid creating a new workspace just to move all the stuff over from the existing workspace.

**Comment by jaysoo** [comment]
Hi @gperdomor, thanks for reporting this issue. As @leosvelperez mentioned, using `--preset=apps` is the way to go to create an empty workspace in the previous style.

```
npx create-nx-workspace --preset=apps
```

The stacks, such as Angular, React, Vue, Node, continue to function like they did prior to v20, and any existing "integrated" workspaces will continue to function the same way in the future. If one of these stacks match what you are using, it is recommended for you to choose them over `None`.

Here's a longer explanation of the error.

Previous to v20, when you chose `None` (i.e. empty workspace) we prompted for package-based vs integrated monorepo. The vast majority of users did not understand this distinction. On top of that, the "integrated" style doesn't use standard features that most modern JS monorepos (Nx or not) are using, such as npm/yarn/pnpm/bun workspaces and TS project references. For those two reasons, we decided, in v20, to create new JS monorepos using the modern setup, which brings them in line with latest best practices, as well as performance improvements.

The `@nx/js` plugin is fully compatible with the new setup, and continues to support the "integrated" style in v20 and into the future. The other plugins, such as `@nx/angular` and `@nx/react` were never fully compatible with the new setup (again, what we used to call "package-based"), and they do cause issues later down the line. It was therefore decided to temporarily guard against them rather than let users discover the issues later on. I want to reiterate that they were always incompatible prior to v20, we're just failing earlier now.

As for the timeline for compatibility for other stacks/plugins, we're actively working on that now. We're targeting v20.1 to bring full compatibility, which is around late October/early November.

**Comment by jaysoo** [comment]
@eduard-malakhov 

> Thank you for your the explanation. What would be the correct way to disable the ts preset in an existing workspace? I'd rather avoid creating a new workspace just to move all the stuff over from the existing workspace.


If the workspace is new, we recommend you re-create it using `--preset=apps`. However, if you need to convert the workspace back to the prevoius, "integrated", style, then you need to:

1. Remove the usage of npm/yarn/pnpm/bun workspaces (i.e. delete `workspaces` entry in `package.json` or remove `pnpm-workspaces.yaml` file)
2. Remove `references` entry in the root `tsconfig.json`

Again, if you can we recommend that you use `--preset=apps` in a new workspace. You can also `nx import` your projects from the existing workspace to the new one.

**Comment by jaysoo** [comment]
@madelk This issue in unrelated, and seems like the `@ns3/nx-serverless` plugin is using an API that was deprecated back in v17, and removed in v20. I would create an issue on their repo.

`context.workspaces.projects` is the problem here, and it was been replaced by `context.projectsConfigurations.projects` in v17.

**Comment by gperdomor** [comment]
Thank you @jaysoo for the clarification... I'm planning to build a component library with Lit so I think I can continue with v20, and create the storybook later when v20.1 is available 👍🏻

**Comment by o-glethorpe** [comment]
How can you guys push a new update and have everything stop working like that? I understand if it happens with third-party plugins, but having this happen with plugins maintained by the NX team itself is unacceptable. 
Everyone who creates a new workspace has to waste time coming down here to find this hack.

**Comment by vzsolt1981** [comment]
> Hi [@gperdomor](https://github.com/gperdomor), thanks for reporting this issue. As [@leosvelperez](https://github.com/leosvelperez) mentioned, using `--preset=apps` is the way to go to create an empty workspace in the previous style.
> 
> ```
> npx create-nx-workspace --preset=apps
> ```
> 
> The stacks, such as Angular, React, Vue, Node, continue to function like they did prior to v20, and any existing "integrated" workspaces will continue to function the same way in the future. If one of these stacks match what you are using, it is recommended for you to choose them over `None`.
> 
> Here's a longer explanation of the error.
> 
> Previous to v20, when you chose `None` (i.e. empty workspace) we prompted for package-based vs integrated monorepo. The vast majority of users did not understand this distinction. On top of that, the "integrated" style doesn't use standard features that most modern JS monorepos (Nx or not) are using, such as npm/yarn/pnpm/bun workspaces and TS project references. For those two reasons, we decided, in v20, to create new JS monorepos using the modern setup, which brings them in line with latest best practices, as well as performance improvements.
> 
> The `@nx/js` plugin is fully compatible with the new setup, and continues to support the "integrated" style in v20 and into the future. The other plugins, such as `@nx/angular` and `@nx/react` were never fully compatible with the new setup (again, what we used to call "package-based"), and they do cause issues later down the line. It was therefore decided to temporarily guard against them rather than let users discover the issues later on. I want to reiterate that they were always incompatible prior to v20, we're just failing earlier now.
> 
> As for the timeline for compatibility for other stacks/plugins, we're actively working on that now. We're targeting v20.1 to bring full compatibility, which is around late October/early November.

With v20.1 and nx/angular the issue still persists.

**Comment by gperdomor** [comment]
@vzsolt1981 yes, I saw PR for [react](https://github.com/nrwl/nx/pull/28808) and [plugins](https://github.com/nrwl/nx/pull/28724) packages but are not merged... I need the last one :(...

I know that the folks are working hard to fix all the compatibility issues but it's worth to say that the release process of this 20 version have been a little chaotic... In any case let's hope that everything can be solved as soon as posible 🙏🏻

**Comment by jaysoo** [comment]
The plugins PR has been merged, and the React one should be landing this week. After that will be Angular, Vue, and Node.

**Comment by oshliaer** [comment]
This doesn't seem quick and easy.

How can I know which plugins support which version? And which monorepo downgrade strategy should I choose?

It would be great to have documentation for this.

**Comment by manpreet-compro** [comment]
Also the matrix I think is for Angular at the moment only. Such a matrix can benefit other plugins also. I am looking for the 'Nuxt' versions, currently not able to create new workspace

**Comment by adlion** [comment]
I tired to create a new project and is failing every time I use angular

What happened to this tool?

**Comment by gperdomor** [comment]
@adlion please read this comment… https://github.com/nrwl/nx/issues/28322#issuecomment-2402623507

Angular plugin is not updated to support the new typescript solution/workspace, you need to use the old style in order to use angular the plugin

**Comment by rmarku** [comment]
With #28808 merged, will the next release fix this?

**Comment by datnd35** [comment]
Let's try again with v19.5.6.

**Comment by datnd35** [comment]
Other solution try with v19.5.6.

**Comment by vietbui** [comment]
> With [#28808](https://github.com/nrwl/nx/pull/28808) merged, will the next release fix this?

This still occurs on the latest version 20.2.2 with `@nx/node` plugin:

```
 NX   The "@nx/node" plugin doesn't yet support the existing TypeScript setup
```

**Comment by datnd35** [comment]
I hope they fix it for `@nx/angular` latest version soon.

**Comment by KerickHowlett** [comment]
I believe I've figured out the problem and will proceed to implement a fix for it.

In the meantime, an easy workaround would be to set the `NX_IGNORE_UNSUPPORTED_TS_SETUP` environment variable to `true` when you try to run the offending command(s).

Example:
```sh
NX_IGNORE_UNSUPPORTED_TS_SETUP=true pnpm dlx nx add @nx/angular
```

**Comment by KerickHowlett** [comment]
From what I can tell, what seems to be throwing this error is the following (assuming you're starting off with an empty workspace using the `None` preset):
- The following checks are failing for the `./tsconfig.json` file:
  - The `files` property being set with a non-empty array.
  - The `references` property being set with a non-empty array.
- The following checks are failing for the `./tsconfig.base.json` file:
  - The `compilerOptions.declaration` boolean property is set to `false`.
  - There are no other defined properties in addition to `compilerOptions` (e.g., `include`).

Just out of curiosity, why are these needed for the new Typescript Solutions Setup?

**Comment by danielritter** [comment]
Issue still here with v20.4.0

**Comment by danielritter** [comment]
Prefixing commands with `NX_IGNORE_UNSUPPORTED_TS_SETUP=true`, e.g. `NX_IGNORE_UNSUPPORTED_TS_SETUP=true nx add @nx/angular` works

**Comment by ChazUK** [comment]
Looks like react, next and nest are all available. It's a shame that nrwl started nx focusing on Angular tooling, and once again it feels like Angular is being left by the wayside in favour of React.

```
% pnpm dlx create-nx-workspace@latest nx-monorepo --pm pnpm
% pnpm dlx nx add @nx/react

✔ Installing @nx/react@20.4.0...
⠧ Initializing @nx/react...Scope: all 2 workspace projects
 WARN  1 deprecated subdependencies found: acorn-import-assertions@1.9.0
Already up to date
Progress: resolved 1157, reused 1052, downloaded 0, added 0, done

dependencies:
+ react 18.3.1 (19.0.0 is available)
+ react-dom 18.3.1 (19.0.0 is available)

Done in 1.7s
✔ Initializing @nx/react...

 NX   Package @nx/react added successfully.

% pnpm dlx nx add @nx/next 

✔ Installing @nx/next@20.4.0...
⠇ Initializing @nx/next...Scope: all 2 workspace projects
 WARN  1 deprecated subdependencies found: acorn-import-assertions@1.9.0
Packages: +3 -1
+++-
Progress: resolved 1197, reused 1068, downloaded 0, added 2, done

dependencies:
+ next 15.1.6

Done in 2.5s
✔ Initializing @nx/next...

 NX   Package @nx/next added successfully.

% pnpm dlx nx add @nx/nest

✔ Installing @nx/nest@20.4.0...
⠋ Initializing @nx/nest...Scope: all 2 workspace projects
 WARN  3 deprecated subdependencies found: acorn-import-assertions@1.9.0, glob@7.2.3, inflight@1.0.6
Packages: +22 -6
++++++++++++++++++++++------
Progress: resolved 1321, reused 1192, downloaded 0, added 13, done

devDependencies:
+ @nestjs/schematics 10.2.3 (11.0.0 is available)

Done in 2.1s
✔ Initializing @nx/nest...

 NX   Package @nx/nest added successfully.

% pnpm dlx nx add @nx/angular                              

✔ Installing @nx/angular@20.4.0...
⠋ Initializing @nx/angular...
 NX   The "@nx/angular" plugin doesn't yet support the existing TypeScript setup

We're working hard to support the existing TypeScript setup with the "@nx/angular" plugin. We'll soon release a new version of Nx with support for it.

✖ Initializing @nx/angular...


 NX   Failed to initialize @nx/angular

  - The "@nx/angular" plugin doesn't yet support the existing TypeScript setup. See the error above.
    Error: The "@nx/angular" plugin doesn't yet support the existing TypeScript setup. See the error above.
```

**Comment by danielritter** [comment]
And so that the angular application compiles correclty again the following needs to be set in `tsconfig.base.json` or in the angular app `tsconfig.json` under `compilerOptions`:

```
	"compilerOptions": {
              "composite": false,
              "emitDeclarationOnly": false,
              "declaration": false,
              "moduleResolution": "node",
              "module": "esnext",
              "lib": ["es2022", "dom"]
              ...
	},
 ```

Without that it does not compile.

**Comment by ChazUK** [comment]
@danielritter One of the reasons I want to use Nx is to manage tsconfigs for me, I've been on too many projects where tsconfig has been tweaked with no explanation, then 6 months down the line something is broken.

**Comment by damiansire** [comment]
![Image](https://github.com/user-attachments/assets/7423eeb4-3267-45d5-981d-250c79bf5d66)

**Comment by damiansire** [comment]
![Image](https://github.com/user-attachments/assets/3cd34aa5-e0f4-4ad4-8b36-68f4bd4f916e)

**Comment by leosvelperez** [comment]
Hey all!

I'm going to close this in favor of the more specific https://github.com/nrwl/nx/issues/29940. Angular is the only stack that still doesn't support the TS project references, and that issue is specifically to track Angular's support. Please, make sure to read https://github.com/nrwl/nx/issues/29940#issuecomment-3003758111 for the last update and follow the issue for future updates regarding this.
